File: Haskell/2022 Julio/01_Hecho/BinaryTree.hs
-------------------------------------------------------------------------------
-- Student's name:
-- Student's group:
-- Identity number (DNI if Spanish/passport if Erasmus):
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module BinaryTree
  ( BinaryTree
  , empty
  , isEmpty
  , insert
  , mkBST
  -- | todo
  , subTreesInRange -- EXERCISE 2
  ) where

data BinaryTree a = Empty
                  | Node a (BinaryTree a) (BinaryTree a)
                  deriving Show


-- ESCRIBE TU SOLUCIÓN DEBAJO ----------------------------------------------
-- WRITE YOUR SOLUTION BELOW  ----------------------------------------------
-- EXERCISE 2

subTreesInRange :: Ord a => BinaryTree a -> a -> a -> Integer
subTreesInRange Empty min max = 0
subTreesInRange (Node x izq der) min max = sol
  where
    lista = recorrerArbol (Node x izq der)
    solmala = length (intervalo min max lista [])
    sol = intInteger solmala

intInteger :: Int -> Integer
intInteger x = toInteger x

recorrerArbol :: Ord a => BinaryTree a -> [a]
recorrerArbol Empty = []
recorrerArbol (Node x izq der) = [x] ++ recorrerArbol izq ++ recorrerArbol der

intervalo :: Ord a => a -> a -> [a] -> [a] -> [a]
intervalo min max [] sol = sol
intervalo min max (x:xs) sol
  | x >= min && x <= max = intervalo min max xs ([x]++sol)
  | otherwise = intervalo min max xs sol

-- | NO MODIFICAR A PARTIR DE AQUÍ --------------------------------------------
-- | DO NOT MODIFY CODE BELOW      --------------------------------------------

empty :: BinaryTree a
empty  = Empty

isEmpty :: BinaryTree a -> Bool
isEmpty Empty = True
isEmpty _     = False

insert :: Ord a => a -> BinaryTree a -> BinaryTree a
insert x' Empty  =  Node x' Empty Empty
insert x' (Node x lt rt)
    | x'<x       = Node x (insert x' lt) rt
    | x'>x       = Node x lt (insert x' rt)
    | otherwise  = Node x' lt rt

mkBST :: Ord a => [a] -> BinaryTree a
mkBST xs  = foldl (flip insert) empty xs

File: Java/2022 Julio/01_Hecho/src/exercises/BinaryTree.java
/* ------------------------------------------------------------------------------
   -- Student's name:
   -- Student's group:
   -- Identity number (DNI if Spanish/passport if Erasmus):
   --
   -- Data Structures. Grado en Informática. UMA.
   -------------------------------------------------------------------------------
*/

package exercises;
//import dataStructures.tuple.Tuple2;

import java.util.ArrayList;

public class BinaryTree<T extends Comparable<? super T>> {

    private static class Node<E> {
        E value;
        Node<E> left;
        Node<E> right;

        public Node(E value, Node<E> left, Node<E> right) {
            this.value = value;
            this.left = left;
            this.right = right;
        }

        public Node(E value) {
            this(value, null, null);
        }
    }

    private Node<T> root;

    public BinaryTree() {
        root = null;
    }

    public boolean isEmpty() {
        return root == null;
    }

    public void insertBST(T value) {
        root = insertBSTRec(root, value);
    }

    private Node<T> insertBSTRec(Node<T> node, T elem) {
        if (node == null) {
            node = new Node<>(elem);
        } else if (elem.compareTo(node.value) < 0)
            node.left = insertBSTRec(node.left, elem);
        else if (elem.compareTo(node.value) > 0)
            node.right = insertBSTRec(node.right, elem);
        else
            node.value = elem;
        return node;
    }

    /**
     * Returns representation of tree as a String.
     */
    @Override
    public String toString() {
        String className = getClass().getSimpleName();
        return className + "(" + toStringRec(this.root) + ")";
    }

    private static String toStringRec(Node<?> node) {
        return node == null ? "null" : "Node<" + toStringRec(node.left) + ","
                + node.value + "," + toStringRec(node.right) + ">";
    }

// -- ESCRIBE TU SOLUCIÓN DEBAJO ----------------------------------------------
// -- WRITE YOUR SOLUTION BELOW  ----------------------------------------------
// -- EXERCISE 2

    public int subTreesInRange(T min, T max) {
        //TODO

        ArrayList<T> l = new ArrayList<>();
        recorrerArbol(root, l);

        int contador = 0;

        for(int i = 0; i<l.size(); i++)
        {
            if (l.get(i).compareTo(min) >= 0 && l.get(i).compareTo(max) <= 0 )
            {
                contador++;
            }
        }

        return contador;
    }

    private void recorrerArbol (Node<T> nodo,  ArrayList<T> l)
    {

        l.add(nodo.value);

        if(nodo.left != null)
        {
            recorrerArbol (nodo.left, l);
        }

        if(nodo.right != null)
        {
            recorrerArbol (nodo.right, l);
        }


    }
}
